libmultilabel/linear/linear.py

```diff
@@ -408,7 +404,61 @@ def cross_validate(y: np.ndarray,
         w = do_train(y[train_idx], x[train_idx], options)
         pred[val_idx] = (x[val_idx] * w).A1 > 0
 
-    return fmeasure(y, 2*pred - 1)
+    return 2*pred - 1
+
+
+def train_cost_sensitive_micro(y: sparse.csr_matrix, x: sparse.csr_matrix, options: str):
+    """Trains a linear model for multilabel data using a one-vs-rest strategy
+    and cross-validation to pick an optimal asymmetric misclassification cost
+    for Micro-F1.
+    Outperforms train_1vsrest in most aspects at the cost of higher
+    time complexity.
+    See user guide for more details.
+
+    Args:
+        y (sparse.csr_matrix): A 0/1 matrix with dimensions number of instances * number of classes.
+        x (sparse.csr_matrix): A matrix with dimensions number of instances * number of features.
+        options (str): The option string passed to liblinear.
+
+    Returns:
+        A model which can be used in predict_values.
+    """
+    # Follows the MATLAB implementation at https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/multilabel/
+    x, options, bias = prepare_options(x, options)
+
+    y = y.tocsc()
+    num_class = y.shape[1]
+    num_feature = x.shape[1]
+    weights = np.zeros((num_feature, num_class), order='F')
+
+    l = y.shape[0]
+    perm = np.random.permutation(l)
+    param_space = [1, 1.33, 1.8, 2.5, 3.67, 6, 13]
+    bestScore = -np.Inf
+    for a in param_space:
+        tp = fn = fp = 0
+        for i in range(num_class):
+            yi = y[:, i].toarray().reshape(-1)
+            yi = 2*yi - 1
+
+            cv_options = f'{options} -w1 {a}'
+            pred = cross_validate(yi, x, cv_options, perm)
+            tp = tp + np.sum(np.logical_and(yi == 1, pred == 1))
+            fn = fn + np.sum(np.logical_and(yi == 1, pred == -1))
+            fp = fp + np.sum(np.logical_and(yi == -1, pred == 1))
+
+        score = 2*tp / (2*tp + fn + fp)
+        if bestScore < score:
+            bestScore = score
+            bestA = a
+
+    final_options = f'{options} -w1 {bestA}'
+    for i in range(num_class):
+        yi = y[:, i].toarray().reshape(-1)
+        yi = 2*yi - 1
+        weights[:, i] = do_train(yi, x, final_options).ravel()
+
```

Comment:
the way to write 458-459 is different from 341-342 and 108-109. Need to be consistent